### DISCUSSION TITLE
Best Dialog practices

### DISCUSSION THREAD

**Discussion by Dremora** [ariakit/ariakit#3412] [Questions] [answered]
I've been trying to wrap my head around how `Dialog`'s `open`, `onClose` and `store` props work in various combinations, and I can't seem to grasp it. Neither can anyone in my team (who are relying on me for my Ariakit knowledge, which I'm failing to provide in this case). 

I'm trying to decide on the best pattern to use the Dialog component and enforce it within our project, and the following questions arise:

- why `open` prop behaves differently to `open` passed to the store? For example, in regards to the [onClose](https://ariakit.org/reference/dialog#onclose) prop:

> It's important to note that this event only fires when the dialog store's [open](https://ariakit.org/reference/use-dialog-store#open) state is set to false. If the controlled [open](https://ariakit.org/reference/dialog#open) prop value changes, or if the dialog's visibility is altered in any other way (such as unmounting the dialog without adjusting the open state), this event won't be triggered.

- same applies to `unmountOnHide` prop when used with vs without a store (#3403), are there any other un-obvious differences?

- Again in relation to #3403, what does it mean for a component not to have a store? Won't it simply not create an internal one and thus should behave in the exact same way as a component with default store?

- why do we have both `onClose` (on the Dialog) and `setOpen` (on the dialog store)? Both nearly overlap in functionality, but there are subtle differences (e.g. #3402). And I can't predict what happens when both are provided.

- With all of these options, what is the recommended approach? Especially considering that Ariakit's API has been evolving since I've started using it (over 1.5 years), and the best practices now might be different. I know that there's tons of docs with examples on the website, but there's no consistency in terms of which approach to use and how to combine them.

**Comment by diegohaz** [comment, marked as answer]
TLDR: Component providers and stores aren't going anywhere. They're used internally to create higher-level APIs. It's perfectly fine to continue using them.

> * why `open` prop behaves differently to `open` passed to the store? For example, in regards to the [onClose](https://ariakit.org/reference/dialog#onclose) prop:
> 
> > It's important to note that this event only fires when the dialog store's [open](https://ariakit.org/reference/use-dialog-store#open) state is set to false. If the controlled [open](https://ariakit.org/reference/dialog#open) prop value changes, or if the dialog's visibility is altered in any other way (such as unmounting the dialog without adjusting the open state), this event won't be triggered.

The `open` prop, when passed to either the component or the store, should work the same way. If the `open` prop gets updated externally, neither the `onClose` nor the `setOpen` callbacks will be invoked. Similarly, the `setOpen` callback won't be called if the component that instantiated the store (the one who calls `useDialogStore`) is unmounted. If you spot any discrepancies in their behavior, please don't hesitate to open an issue. I'm also open to suggestions on how to enhance the clarity of this description.

> * same applies to `unmountOnHide` prop when used with vs without a store ([unmountOnHide doesn't respect closing animation when Dialog is used without explicit store #3403](https://github.com/ariakit/ariakit/issues/3403)), are there any other un-obvious differences?
>
> * Again in relation to [unmountOnHide doesn't respect closing animation when Dialog is used without explicit store #3403](https://github.com/ariakit/ariakit/issues/3403), what does it mean for a component not to have a store? Won't it simply not create an internal one and thus should behave in the exact same way as a component with default store?

`unmountOnHide` on animated dialogs is the only difference I'm aware of.

This is just an implementation detail, but `Dialog` does have an internal store. However, the component could get completely unmounted when you use the `unmountOnHide` prop, which also wipes out the store. We might consider supporting this in the future by also creating the store in the intermediate component that determines if the underlying `Dialog` will be mounted or not. For now, the best approach is to use a store or wrap the `Dialog` with `DialogProvider`.

> * why do we have both `onClose` (on the Dialog) and `setOpen` (on the dialog store)? Both nearly overlap in functionality, but there are subtle differences (e.g. [Dialog's open state is not fully controllable #3402](https://github.com/ariakit/ariakit/issues/3402)). And I can't predict what happens when both are provided.

Could you clarify the differences you noticed between `onClose` and `setOpen`? Actually, `onClose` uses `setOpen` internally, so their behavior should be identical, except that `onClose` gets a cancelable event as an argument. If you've observed any discrepancies, it could potentially be a bug.

The `open` and `onClose` props were added to `Dialog` to align with the native dialog API. They are built on top of stores. This is a higher-level alternative to `DialogProvider` and `useDialogStore`.

> * With all of these options, what is the recommended approach? Especially considering that Ariakit's API has been evolving since I've started using it (over 1.5 years), and the best practices now might be different. I know that there's tons of docs with examples on the website, but there's no consistency in terms of which approach to use and how to combine them.

Typically, you can start with the higher-level APIs (`open`/`onClose`) until you need features only available through lower-level APIs. The API was designed to ensure this progression is as smooth as possible. This holds true for most Ariakit components. The majority of users will stick to this level, as they don't need advanced features.

However, component providers and stores aren't going anywhere. They're used internally to supply these higher-level APIs. It's perfectly fine to use them.

**Reply by Dremora** [reply]
Thanks for the detailed response! My main takeaway is that I should aim to use higher-level APIs unless I need stores. Due to the behaviour of `unmountOnHide`, I need to use stores.

> Could you clarify the differences you noticed between onClose and setOpen?

The question was not on the functional difference, which I understand, but rather why have both. Now I understand that `onClose` is a higher-level API (and as you said, it's mirroring dialog's native API). We can also combine both callbacks.

> The open prop, when passed to either the component or the store, should work the same way. If the open prop gets updated externally, neither the onClose nor the setOpen callbacks will be invoked. Similarly, the setOpen callback won't be called if the component that instantiated the store (the one who calls useDialogStore) is unmounted. If you spot any discrepancies in their behavior, please don't hesitate to open an issue. I'm also open to suggestions on how to enhance the clarity of this description.

This is what the docs say in regards to `open` and `onClose`:

> It's important to note that this event only fires when the dialog store's [open](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) state is set to false. If the controlled [open](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) prop value changes, or if the dialog's visibility is altered in any other way (such as unmounting the dialog without adjusting the open state), this event won't be triggered.

My understanding is that `open` only refers to the component's prop, and not to the store's state. And it seems this is how it currently works. As you can see in [this example](https://stackblitz.com/edit/lzpywb?file=dialog%2Fstyle.css,dialog%2Findex.tsx&theme=dark), trying to close the modal via `onClose` event re-opens the dialog immediately. It seems the event is getting triggered multiple times. I'm surprised this doesn't cause an infinite loop.

**Reply by Dremora** [reply]
I've raised a bug: #3488.
